### ISSUE TITLE
exec: add support for ON expression for the joiners

### ISSUE THREAD

**Issue by yuzefovich** [cockroachdb/cockroach#38018]
Currently, vectorized engine doesn't support `ON expression` on:
 - [ ] hash joiner
 - [ ] merge joiner

**Comment by yuzefovich** [comment]
I wonder how important getting this done is for 20.2 release. cc @asubiotto @jordanlewis thoughts?

@awoods187 do we happen to have a telemetry on this kind of thing (i.e. whether an ON expression is used for a particular join type)? Probably not, but I thought I'd ask.

**Comment by asubiotto** [comment]
It's a good question. I think completing long-tail work items for vectorized support is currently low priority. It would be great if we can get it done, but I don't think we should drop anything to do it unless we get customer data that would make us reasess. I would argue that this work is closely related to the "add remaining datatypes to vectorize" milestone item.

**Comment by yuzefovich** [comment]
I agree in general about the long-tail work items, but I wonder whether this particular issue needs to be singled out. My reasoning is that joins are one of the operation types that gain the most performance improvements from the vectorization, and if many joins do use ON expression, we're currently wrapping row-execution processors into the plan, possibly leaving most of the performance boost on the table.

In my opinion this issue is of higher priority that remaining datatypes because we already have support for the most commonly used types that can be optimized and we have reasonable datum-backed vector fallback. For ON expression the fallback to rowexec doesn't seem equivalent. That said, I don't have a feeling of how often users actually have queries that end up with ON expression.

**Comment by yuzefovich** [comment]
Oh, I just thought about this issue a little bit more, and it might be no longer as tricky as it was when I looked into this initially. Previously, we had an assumption built into the hash and merge joiners that only "necessary" columns need to be output, and this assumption forced us to "push" the ON expression filtering logic right into the "heart" of the joiners.

However, we have removed that assumption and now output all of the columns (and plan a simple projection on top of the joiner if necessary), so I think now this issue becomes quite approachable. The joiners will still need to be aware of the ON expression filtering (e.g. for LEFT SEMI join having a single row from the right match the probing row on the left might no longer be sufficient if the ON condition is not satisfied, etc). I think that this might be a good issue for @helenmhe to work on during her internship.
